app/services/activitypub/process_status_update_service.rb

```diff
@@ -294,21 +293,35 @@ def update_quote!
         quote = Quote.create(status: @status, approval_uri: approval_uri)
         @quote_changed = true
       end
-    end
 
-    if quote.present?
-      begin
-        quote.save
-        ActivityPub::VerifyQuoteService.new.call(quote, fetchable_quoted_uri: quote_uri, request_id: @request_id)
-      rescue Mastodon::UnexpectedResponseError, *Mastodon::HTTP_CONNECTION_ERRORS
-        ActivityPub::RefetchAndVerifyQuoteWorker.perform_in(rand(30..600).seconds, quote.id, quote_uri, { 'request_id' => @request_id })
-      end
+      quote.save
+
+      fetch_and_verify_quote!(quote, quote_uri)
     elsif @status.quote.present?
       @status.quote.destroy!
       @quote_changed = true
     end
   end
 
+  def fetch_and_verify_quote!(quote, quote_uri)
+    ActivityPub::VerifyQuoteService.new.call(quote, fetchable_quoted_uri: quote_uri, prefetched_quoted_object: safe_prefetched_quoted_object, request_id: @request_id)
+  rescue Mastodon::UnexpectedResponseError, *Mastodon::HTTP_CONNECTION_ERRORS
+    ActivityPub::RefetchAndVerifyQuoteWorker.perform_in(rand(30..600).seconds, quote.id, quote_uri, { 'request_id' => @request_id })
+  end
+
+  def safe_prefetched_quoted_object
+    object = @status_parser.quoted_object
+    return unless object.is_a?(Hash)
+
+    # NOTE: Replacing the object's context by that of the parent activity is
+    # not sound, but it's consistent with the rest of the codebase
+    object = object.merge({ '@context' => @activity_json['@context'] })
+
+    return if value_or_id(first_of_value(object['attributedTo'])) != @account.uri || non_matching_uri_hosts?(@account.uri, object['id'])
+
+    object
+  end
```

Comment:
I am not sure about repeating this method from `ActivityPub::Activity::Create` above here (or maybe it is the other way around :slightly_smiling_face:).

DRY is not an end in and of itself, but this logic here is somewhat sensitive, so I think there would be a real benefit to have this consistent in a central location.

At the same time I realize it is not at all clear what that location could be :frowning_face: 
`JsonLdHelper` would be a somewhat obvious choice as both classes already include it and this method calls other methods from that module, but of course this is not really related to JSON-LD in any way. Maybe it could be worth it to add a new module, `QuoteHelper` or even `QuoteVerificationHelper`, for this?

Reply:
I refactored it in a slightly more generic `safe_prefetched_embed` helper in `JsonLdHelpers`.

Reply:
Very nice!